### PR TITLE
Render Markdown in search results

### LINKED ISSUE #32040: Rustdoc search results ignore markdown formatting
Using Rust 1.7, a rustdoc search result shown in the browser (as a consequence of pressing the `S` key and entering a search term) displays raw Markdown, not formatted text. For example, take the standard library documentation, press `S`, then enter the search term `trait: Neg`, and the search result will show this text verbatim:

```
The `Neg` trait is used to specify the functionality of unary `-`.
```

Whereas, it should be formatted—i.e., no backticks.

### PR DESCRIPTION
**Pull request by camelid**
Fixes #32040.

Previously Markdown documentation was not rendered to HTML for search results,
which led to the output not being very readable, particularly for inline code.
This PR fixes that by rendering Markdown to HTML with the help of pulldown-cmark
(the library rustdoc uses to parse Markdown for the main text of documentation).
However, the text for the title attribute (the text shown when you hover over an
element) still uses the plain-text rendering since it is displayed in browsers
as plain-text.

Only these styles will be rendered; everything else is stripped away:

* *italics*
* **bold**
* `inline code`

### CODE REVIEW

**Comment by camelid** [comment]
Here's what it looks like:

image

Still some bugs to fix, like the brackets in the code links. Interestingly, links are rendered correctly with plain text (i.e., rendered as just the link text), but not with inline code. Also, some of the summaries have an extra space before the ellipsis.

**Comment by camelid** [comment]
Also the summary for [`Command::arg0()`](https://doc.rust-lang.org/std/process/struct.Command.html#method.arg0) isn't being rendered for some reason. Maybe it's because it's platform-specific? (It's not being rendered in the official docs either though, so not because of this.)

**Comment by jyn514** [comment]
cc @euclio , you know this code better than I do.

**Comment by jyn514** [comment]
@camelid have you considered the issues ollie mentioned in https://github.com/rust-lang/rust/pull/68699#issuecomment-581079387 ?

**Comment by camelid** [comment]
Ah, I was trying to figure out where the alt-text was used. I'll add back the plain summary function in addition, and use that for alt-text.

**Comment by jyn514** [comment]
cc @GuillaumeGomez , this replaces https://github.com/rust-lang/rust/pull/68699. I'm not yet sure how it differs from that PR, although I think it does go through pulldown instead of re-parsing markdown in javascript.

**Comment by camelid** [comment]
Yes, this approach is to render the HTML using pulldown ahead of time.

**Comment by camelid** [comment]
Hmm, it's still rendering alt-text as HTML...

**Comment by camelid** [comment]
Rebased.

File: src/librustdoc/html/markdown.rs
Comment (jyn514) [inline src/librustdoc/html/markdown.rs:1043]:
```suggestion
/// Returns a tuple of the rendered HTML string and whether the output was shortened.
```

File: src/librustdoc/html/static/main.js
Comment (jyn514) [inline src/librustdoc/html/static/main.js:1550]: This is no longer escaped because it's HTML, right? Can you add a comment saying that?
```suggestion
                              // the description can be HTML, so don't escape it
                              "<span class=\"desc\">" + item.desc +
```
I'm also a little concerned this a change in behavior ... where is the description coming from? What would have been escaped before and isn't now - can you add a test?

File: src/librustdoc/html/render/mod.rs
Comment (jyn514) [inline src/librustdoc/html/render/mod.rs:1582]: This is still rendering as plain-text, right? Why did you change the comment?

File: src/librustdoc/html/markdown.rs
Comment (jyn514) [inline src/librustdoc/html/markdown.rs:1127]: I liked this before, is there a reason to change it? Being called `render` without being in the `render` module seems confusing.

**Review by jyn514 — COMMENTED** [review]
> Ah, I was trying to figure out where the alt-text was used. I'll add back the plain summary function in addition, and use that for alt-text.

Did you fix this? If so, can you add a test?

We discussed in PMs that you're still working on this so I'm not doing a full review for now - let me know when you've got this working and I'll take another look.

File: src/librustdoc/html/markdown.rs
Comment (camelid) [reply src/librustdoc/html/markdown.rs:1127]: I guess not; I just try to have function names be actions, rather than nouns :)

File: src/librustdoc/html/render/mod.rs
Comment (camelid) [reply src/librustdoc/html/render/mod.rs:1582]: Thanks! I think I forgot to update it when I switched back to using plaintext for alt-text.

File: src/librustdoc/html/markdown.rs
Comment (camelid) [reply src/librustdoc/html/markdown.rs]: And I also wanted it be consistent with `render_markdown_summary*`.

File: src/librustdoc/html/markdown.rs
Comment (jyn514) [inline src/librustdoc/html/markdown.rs:1045]: Can you either:
- Move these files from `html/` to `render/`
- Remove `render_` from the start

Having this called 'render' without being in the render module is very confusing.

File: src/librustdoc/html/static/main.js
Comment (camelid) [reply src/librustdoc/html/static/main.js:1550]: Where would I add the test: would it be a unit test or an integration test?

Comment (jyn514) [reply src/librustdoc/html/static/main.js:1550]: This is in JavaScript, so it would go in `src/test/rustdoc-js` (the javascript doesn't have unit tests at all). It should test the change in behavior, which I think is that search results will no longer be escaped? I don't know exactly what that would look like, that's one of the reasons I'd like a test.

File: src/librustdoc/html/static/main.js
Comment (jyn514) [inline src/librustdoc/html/static/main.js:1980]: This seems really fragile ... is there a way to do this in rustdoc itself, where we know what was added by rustdoc and what was added by the user? The issue is that anything we store makes the index take longer to render ... I'm not sure the best approach here.

@GuillaumeGomez do you have ideas?

**Review by jyn514 — COMMENTED** [review]
I don't think I'm the best reviewer for this.

r? @GuillaumeGomez

File: src/librustdoc/html/static/main.js
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js:1980]: It's definitely not the right approach. The good approach is to generate the HTML directly from rustdoc into the search index.

Comment (camelid) [reply src/librustdoc/html/static/main.js:1980]: @GuillaumeGomez What do you mean? The HTML *is* being rendered into the search index; that's why we need this function.

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js:1980]: What I mean is that if there is a `code` attribute in the first line, the transformation into HTML should be done in rustdoc when generating the doc. Nothing should be done on the JS side.

Comment (camelid) [reply src/librustdoc/html/static/main.js:1980]: It is being transformed into HTML in rustdoc. The JS transforms the HTML to plaintext for the alt-text.

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js:1980]: Oh I see! Sorry, I was completely out of it. But then, is it really worth it?

File: src/librustdoc/html/static/main.js
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/main.js:1967]: Just noticed but that's not a JS comment. :p

File: src/librustdoc/html/static/main.js
Comment (camelid) [reply src/librustdoc/html/static/main.js:1980]: That's okay :)

Is what worth it?

File: src/librustdoc/html/static/main.js
Comment (camelid) [reply src/librustdoc/html/static/main.js:1967]: Yeah, I wrote it like a rustdoc comment, but this *is* JavaScript.

File: src/librustdoc/html/static/main.js
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js:1980]: To generate the alt-text with this information. Isn't `DOMElement.innerText` enough? Do we really want to keep things like backquote?

File: src/librustdoc/html/static/main.js
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js:1967]: It's valid, but unexpected. Please use "//" haha.

Comment (camelid) [reply src/librustdoc/html/static/main.js]: I converted it to a JSDoc comment.

File: src/librustdoc/html/static/main.js
Comment (camelid) [reply src/librustdoc/html/static/main.js]: I still don't understand what you mean :/

The JS code previously used the search index to fill the crate list in the sidebar, and used the (plaintext) summary for search result as the alt-text (actually the `title` attribute) on each crate in the list. But now the search result summaries are HTML. However, we can't use HTML in the alt-text since it is rendered literally to the user, i.e. if you hover over an element it will show a tooltip that looks like "This crate has a \<code\>Foo\</code\> struct". It used to look like "This crate has a \`Foo\` struct".

So the solution I went with for now is to still get the alt-text from the search index, but to first pass it through a function to convert it to plaintext. So I think the worst that can happen if this function doesn't work correctly is that the alt-text will look a little funny. Personally I think for now we should go with this approach and then in a separate PR change it so that it doesn't convert the HTML to plaintext in the JS. That way this PR's scope will stay somewhat small.

**P.S.** When I say "alt-text" I mean the text in the tooltip that is shown when you hover over an element.

File: src/librustdoc/html/static/main.js
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js]: :+1:

File: src/librustdoc/html/static/main.js
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js]: So I understood that part correctly. What I mean is that what you're doing is just way too much. :) First: never parse HTML directly, if we really want to do that, then use the DOM api (`childNodes` and equivalents). Second: this is way too much, just get the text using the text you get after setting the HTML to the DOMElement and then use `innerText` which will return the text only and no HTML.

Comment (camelid) [reply src/librustdoc/html/static/main.js]: I think I may understand what you mean: You're suggesting that I use something https://stackoverflow.com/a/21870431 and use the built-in `DOMParser` to parse the HTML and then just use `.innerText` on it? I hadn't thought of that; good idea!

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js]: I just realized that we created the whole table results as a string that we set to the `innerHTML`. That's why what I said didn't make sense (sorry!). Instead of this whole thing, we could instead create each line using `document.createElement` and so on. And then, on the created table entry (the `<tr>`), you can then get its last `<td>.innerText`. No need to parse anything. But you might need to rewrite a bit how we generate the results output. I can do it if you want?

Comment (camelid) [reply src/librustdoc/html/static/main.js]: I'm sorry, but I still don't understand what you are suggesting. We could just use `DOMParser` to parse the HTML and then use `.innerText`, but then inline code would not have backticks around it. But that's not a big deal since this is just alt-text. I don't see how we can have alt-text without stripping the tags of the HTML.

Comment (camelid) [reply src/librustdoc/html/static/main.js]: This parsing is *purely* for the alt-text (`title` attribute), not for what's displayed in the rows. I'm not sure why what you're suggesting is relevant to rendering the alt-text. In a different PR, it may make sense to refactor this code, but I don't see why it's necessary now.

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js]: I'll send a PR to make the change I suggested then, so that then you can use it. It'll be simpler haha.

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js]: I opened https://github.com/rust-lang/rust/pull/78805. You can cherry-pick the commit if you want or just land the PR and then rebase on it, as you prefer.

Comment (camelid) [reply src/librustdoc/html/static/main.js]: Oh, that code is what I thought you meant :)

File: src/librustdoc/html/markdown.rs
Comment (GuillaumeGomez) [inline src/librustdoc/html/markdown.rs:1045]: Why is `length_limit` an `Option` and not just `u16`? It's only used in one place after all...

File: src/librustdoc/html/static/main.js
Comment (jyn514) [inline src/librustdoc/html/static/main.js:1980]: This bit is no longer necessary I think.
```suggestion
            html,
```

Comment (jyn514) [reply src/librustdoc/html/static/main.js:1980]: Oh I take it back - it is necessary if you want to replace code blocks with \`. Not sure why rustdoc is doing that, though.

File: src/librustdoc/html/markdown.rs
Comment (GuillaumeGomez) [inline src/librustdoc/html/markdown.rs:1059]:
```suggestion
                for (i, word) in text.split_whitespace().enumerate() {
```

File: src/librustdoc/html/markdown.rs
Comment (GuillaumeGomez) [inline src/librustdoc/html/markdown.rs:1097]: You forgot to increase the text length. Also: maybe create a function to push text/character and increase the text length at the same time.

File: src/librustdoc/html/static/main.js
Comment (camelid) [reply src/librustdoc/html/static/main.js]: I ended up using something similar to what you suggested, but instead used `DOMParser`, which is designed explicitly for parsing HTML and manipulating it as a DOM element. It also seems to be supported in pretty much every browser.

File: src/librustdoc/html/markdown.rs
Comment (GuillaumeGomez) [inline src/librustdoc/html/markdown.rs:1099]: I think you need to add the `if text_length + word.len() >= length_limit` check here as well.

File: src/librustdoc/html/static/main.js
Comment (camelid) [reply src/librustdoc/html/static/main.js:1980]: It's not strictly necessary, but I think it's better to do it this way, for one since that's how it used to work.

File: src/librustdoc/html/static/main.js
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js:1550]: You can check the results returned by the search function, so including the `doc` field that has been updated here to confirm that it still contains HTML tags (and didn't escape them).

File: src/librustdoc/html/static/main.js
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js:1980]: Well, it *kinda* makes sense. However, we also keep other HTML tags like `<b>` and `<i>`. Should we also "revert" them into markdown too?

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js:1980]: > It's not strictly necessary, but I think it's better to do it this way, for one since it used to use ` here.

My comment still stands. :p

Comment (jyn514) [reply src/librustdoc/html/static/main.js:1980]: I really don't think it makes sense to parse HTML ourselves - we have the same issue as in https://github.com/rust-lang/rust/pull/68699#issuecomment-627465641 that we don't know what was originally html and what was markdown. If we don't want to show rendered text, we shouldn't show it, but I don't think we should try to rewrite some sort of HTML to markdown converter.

Comment (camelid) [reply src/librustdoc/html/static/main.js:1980]: > we have the same issue as in [#68699 (comment)](https://github.com/rust-lang/rust/pull/68699#issuecomment-627465641) that we don't know what was originally html and what was markdown

I do not think that is correct. The example given in https://github.com/rust-lang/rust/pull/68699#issuecomment-627465641 was:

```rust
/// foo `this is a code span` bar \`this isn't\` baz
pub struct Foo;
```

That is stored in the search index as:

```html
foo <code>this is a code span</code> bar `this isn't` baz
```

And then `convertHTMLToPlaintext` converts that into:

```
foo `this is a code span` bar `this isn't` baz
```

Which is *exactly* the same as the current hover text.

-----

That said, I am open to just stripping every HTML tag if we decide that would be better. I just don't think https://github.com/rust-lang/rust/pull/68699#issuecomment-627465641 is relevant.

Comment (jyn514) [reply src/librustdoc/html/static/main.js:1980]: @camelid the test case is the other way:

```rust
/// foo <code>this is a code span</code> in HTML
///
/// `this is markdown` baz
pub struct Foo;
```

There's no way to get back that info after the fact.

Comment (camelid) [reply src/librustdoc/html/static/main.js:1980]: I just copy-pasted the test case from the comment you linked to. The comment doesn't seem to be taking issue with the example you're giving. This is the full comment text:

> No. The summaries stored in the search index don't contain that information. For:
>
> ```rust
> /// foo `this is a code span` bar \`this isn't\` baz
> pub struct Foo;
> ```
>
> the search index contains: 
>
> ```text
> foo `this is a code span` bar `this isn't` baz
> ```
>
> There is no way in JavaScript to know which backticks meant codeblocks and which didn't.
>
> As I've said, to do this right the summaries need to be rendered and stored in the search index as HTML.

It is saying that converting backticks into HTML is a problem because they both mean inline code and literal backticks. Converting in the opposite direction is loss-less.

-----

And, either way, I don't see how converting both `` `code` `` and `<code>code</code>` into `` `code` `` when this is just for alt-text is an issue. If for some reason this is a blocker, I am absolutely fine to just strip all HTML tags.

File: src/librustdoc/html/markdown.rs
Comment (camelid) [reply src/librustdoc/html/markdown.rs:1045]: I wanted to give the ability to render a summary without a length limit in the future.

Comment (GuillaumeGomez) [reply src/librustdoc/html/markdown.rs:1043]: We have to limit it otherwise the search index size will be too huge.

File: src/librustdoc/html/static/main.js
Comment (jyn514) [reply src/librustdoc/html/static/main.js:1980]: I'm not 100% sure about this but I don't think it should block this change. So if @GuillaumeGomez thinks this is good to go I do too.

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js:1980]: I said I didn't. :stuck_out_tongue:

**Comment by camelid** [comment]
Rebased.

**Comment by GuillaumeGomez** [comment]
Based on this [document](https://developer.mozilla.org/fr/docs/Web/API/DOMParser), DOMParser isn't supported on mobiles devices. The best way would still be to go through an HTML element I think:

```js
var x = document.createElement("div");
x.innerHTML = html;
return x.innerText;
```

**Comment by camelid** [comment]
@GuillaumeGomez I'm not sure why you say that it isn't supported on mobile devices. This is what the English version of the page shows:

image

Maybe the French version is out of date?

Note that the only two that don't say for sure supported are iOS Safari and Opera for Android. I checked recently and confirmed that in fact it is supported on iOS Safari and [opened a PR that was merged](https://github.com/mdn/browser-compat-data/pull/7308) to update it. So that just leaves Opera for Android as compatibility-unknown, and I doubt that there is a significant number of people who use Opera for Android, are browsing Rust documentation on their phone, *and* want to see the title text for the crate list sidebar. I still think using DOMParser is the better approach because it is explicitly designed for parsing HTML and it seems to be supported practically everywhere.

**Comment by camelid** [comment]
The French page says

> Nous convertissons les données de compatibilité dans un format JSON. Ce tableau de compatibilité utilise encore l'ancien format car nous n'avons pas encore converti les données qu'il contient. Vous pouvez nous aider en contribuant !

which (according to Google Translate) means

> We convert the compatibility data into a JSON format. This compatibility table still uses the old format because we have not yet converted the data in it. You can help us by contributing!

I think it's saying that the compatibility data is outdated.

**Comment by GuillaumeGomez** [comment]
Weird... In any case, `DOMParser` is `?` for `iOS safari`, meaning that it's very likely that no web browser supports it on iOS.

**Comment by camelid** [comment]
> I checked recently and confirmed that in fact it is supported on iOS Safari and [opened a PR that was merged](https://github.com/mdn/browser-compat-data/pull/7308) to update it.

@GuillaumeGomez ^

**Comment by GuillaumeGomez** [comment]
Why not just go for the code I gave already? It doesn't tell us since when it's supported on iOS and I don't really see much advantage from it...

**Comment by camelid** [comment]
> It doesn't tell us since when it's supported on iOS

What do you mean? If you look at the PR, it says it's been supported since version 9.3, which is around 5 years old at this point.

> I don't really see much advantage from it...

The advantages are:

1. It's a cleaner approach since it's explicitly designed for this sort of use case
2. It's supported on practically every platform

So why *not* use it? I don't want to spend all the time arguing about this though, so if for some reason it is a sticking point I can switch to using `createElement`, but I don't think it is the right approach and I don't see what's wrong with DOMParser.

**Comment by GuillaumeGomez** [comment]
Sorry if I sounded a bit harsh, it's just that I had so many issues with mobile web browsers that I'm very careful about them now. Another source: https://caniuse.com/?search=DOMParser

The support for DOMParser isn't that clear, which is why I'm a bit afraid to be honest. I agree with you, it'd be a lot nicer to use it because it makes its usage more clear, but I'm having cold feet because the support is really unclear. So here what I suggest, if other people from @rust-lang/rustdoc agree with you, then I approve this PR as well (when you have fixed the CI issues and cleaned your commits a bit because you have 22 of them now, which is a bit too much haha). Does that sound reasonable to you?

And again: sorry you had to go through all this, your work is really appreciated. This is a really nice improvement and I can't wait for it to be merged!

**Comment by camelid** [comment]
> Sorry if I sounded a bit harsh, it's just that I had so many issues with mobile web browsers that I'm very careful about them now.

Okay, I understand.

> The support for DOMParser isn't that clear

Here's my reasoning:

1. All but Opera for Android are confirmed to work with DOMParser, and iOS Safari has worked with it for 5 years

2. This is just for the title text on the items in the crate list sidebar, and I don't think you can even view title text on mobile. If a browser doesn't support it, it shouldn't affect the rest of the page.

3. If for some reason there *is* an issue that is reported with this, we can always switch to using `createElement` – i.e., this isn't a big architectural decision

> And again: sorry you had to go through all this, your work is really appreciated. This is a really nice improvement and I can't wait for it to be merged!

Thanks, me too!

**Comment by GuillaumeGomez** [comment]
> This is just for the title text on the items in the crate list sidebar, and I don't think you can even view title text on mobile. If a browser doesn't support it, it shouldn't affect the rest of the page.

Yes, but if JS crashes at runtime, well, no search results. So even if you don't see the title text on mobile, it still has an impact.

**Comment by jyn514** [comment]
> It used to use `split(' ')` combined with pushing a space before
the word if it was after the first word. This seemed to work, although
it likely had potential for bugs. Then, when we switched to using
`split_whitespace`, it stopped working -- I'm actually not quite sure
why.

@camelid `split_whitespace` includes things that aren't spaces, like tabs and newlines.

**Comment by camelid** [comment]
I understand that; I don't understand why the output mostly didn't have spaces between words. It looked like this:

> Thisisasummarywith alinkandcode

**Comment by camelid** [comment]
Squashed in preparation for rebasing (too many conflicts otherwise, and I need to squash anyway).

**Comment by jyn514** [comment]
@GuillaumeGomez the support policy is the last two Chrome versions, Firefox ESR, and the latest version of all other major browsers: https://rust-lang.github.io/rfcs/1985-tiered-browser-support.html#supported-browsers. I don't think we should block this on support for a 5-year old browser.

**Comment by GuillaumeGomez** [comment]
We currently support down to IE 10 iirc. I don't see why we should drop this for 1 line of JS that can be transformed into 3 lines without problems...

**Comment by camelid** [comment]
We *wouldn't* be dropping support for IE 10 -- support for `parseFromString(..., "text/html")` was added in that version. That said, I don't want to argue about this, so I'll switch to using `.innerHTML`.

**Comment by GuillaumeGomez** [comment]
Thank you.

File: src/librustdoc/html/static/main.js
Comment (jyn514) [inline src/librustdoc/html/static/main.js:2043]: This drops the `<code>`handling, was that intentional?

Comment (camelid) [reply src/librustdoc/html/static/main.js]: No it was not, thank you! I actually noticed that I had removed the handling by mistake, but then I forgot to add it back 🤦

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js]: Well, I personally think it's better this way but it's up to you. :)

**Comment by camelid** [comment]
By the way, I think this is ready to be reviewed! There are still some unresolved questions:

https://github.com/rust-lang/rust/pull/77686#discussion_r511495209
https://github.com/rust-lang/rust/pull/77686#discussion_r518982482
https://github.com/rust-lang/rust/pull/77686#discussion_r518983038

but I would appreciate it if we could go over those again since it's been so long.

**Comment by jyn514** [comment]
> https://github.com/rust-lang/rust/pull/77686#discussion_r511495209

I think this is still waiting on a test.

> https://github.com/rust-lang/rust/pull/77686#discussion_r518982482

I agree with Guillaume, I don't think the change makes sense.

> https://github.com/rust-lang/rust/pull/77686#discussion_r518983038

Converting `<code></code>` unconditionally seems ok to me - I'm still not sure rustdoc should be doing that, but this isn't the place to change it.

File: src/librustdoc/html/markdown.rs
Comment (camelid) [reply src/librustdoc/html/markdown.rs]: Done.

File: src/librustdoc/html/static/main.js
Comment (camelid) [reply src/librustdoc/html/static/main.js:1613]: Done! Finally.

**Comment by camelid** [comment]
Unless Guillaume doesn't agree with the discussion in https://github.com/rust-lang/rust/pull/77686#discussion_r518983038, this should be ready to merge!

**Comment by camelid** [comment]
Unfortunately this is broken with intra-doc links. Going to look into it tomorrow.

File: src/test/rustdoc-js/summaries.js
Comment (GuillaumeGomez) [inline src/test/rustdoc-js/summaries.js:1]: You can make multiple queries at once by using an array (then `EXPECTED` has to be an array as well).

Take a look at `doc-alias.js` if you're not sure how it should look like.

File: src/librustdoc/html/static/main.js
Comment (GuillaumeGomez) [inline src/librustdoc/html/static/main.js:2033]:
```suggestion
     *   * Replace "<code>foo</code>" with "`foo`"
```

;)

File: src/librustdoc/html/markdown.rs
Comment (GuillaumeGomez) [inline src/librustdoc/html/markdown.rs:1116]: Just thought about this but since `markdown_summary_with_limit` is only used in this case, shouldn't we do this directly inside it? And maybe even remove this function? Not sure what you think about this.

Maybe you have an opinion too @jyn514 ?

Comment (jyn514) [reply src/librustdoc/html/markdown.rs:1116]: Yeah, I think it makes sense to inline `markdown_summary_with_limit` into here.

Comment (GuillaumeGomez) [reply src/librustdoc/html/markdown.rs:1116]: What do you think @camelid ?

File: src/test/rustdoc-js/summaries.rs
Comment (GuillaumeGomez) [inline src/test/rustdoc-js/summaries.rs:12]: So now just waiting for the tests on `Sidebar` and `Sidebar2`. ;)

File: src/librustdoc/html/markdown.rs
Comment (camelid) [reply src/librustdoc/html/markdown.rs:1116]: Well I think it's better to keep them separate, for two reasons:

1. Just because we only use `markdown_summary_with_limit` in this case doesn't mean we will never use it elsewhere

2. I think it makes the code easier to understand when there's more of a separation of concerns

But I'm also really eager to get this done, so I'm willing to inline it if you disagree with my reasoning.

Comment (GuillaumeGomez) [reply src/librustdoc/html/markdown.rs:1116]: It's not a blocking point for me, more like an unnecessary complexity. So unless @jyn514 wants to wait for it to be inlined, we can always come back to it later on.

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Looks good to me, thanks a lot! Just waiting for the opinion of @jyn514 about the inlining and then r=me,jyn514!

**Comment by camelid** [comment]
Rebased.

**Comment by GuillaumeGomez** [comment]
CI failed. ;)

**Comment by camelid** [comment]
Can you let me know before this is merged so I can squash? Or I can squash now if you would like.

**Comment by GuillaumeGomez** [comment]
7 commits doesn't seem excessive to me so unless someone else wants to, I think you're fine. ;)

**Comment by camelid** [comment]
Oh... well, they're just going back and forth on implementation, so I thought it might be cleaner to squash. But I'm fine with either.

**Comment by jyn514** [comment]
Seems fine to me.

@bors r=GuillaumeGomez

**Comment by Mark-Simulacrum** [comment]
@bors squash

I think squashing would be good, and we should have automated squashing now I believe...

**Comment by camelid** [comment]
Hopefully it won't ruin the commit message...

**Comment by camelid** [comment]
@Mark-Simulacrum Wait, but this will make the squashed commit show as authored by bors.

@bors squash-

**Comment by Mark-Simulacrum** [comment]
I am confused -- my understanding is that the functionality in bors does not produce that result? It'll set the committer to bors most likely but that seems fine?

**Comment by camelid** [comment]
I'm not sure, but I'd rather squash by hand if you want me to squash.

**Comment by Mark-Simulacrum** [comment]
I would personally like to at least get rid of the "Fix typo" commit, I don't care strongly about others though.

**Comment by camelid** [comment]
I squashed in the "Fix typo" commit. @Mark-Simulacrum could you `r=GuillaumeGomez`?

**Comment by Mark-Simulacrum** [comment]
I guess 376507f47b5f16c0a9c210005de2bcdb270d8920 is also a typo but I'm out of energy here so @bors r=GuillaumeGomez delegate+

**Comment by camelid** [comment]
If it's a big deal, I can squash more away, but I would really like to get this merged.

**Comment by GuillaumeGomez** [comment]
> If it's a big deal, I can squash more away, but I would really like to get this merged.

Don't worry it wasn't a big issue. Thanks a lot for working on this!

**Comment by camelid** [comment]
Yay! Excited to look at the docs for the next nightly :)

**Pull request merged**
